bucket_api.go

```diff
@@ -339,40 +337,20 @@ func (bucket *Bucket) ExpireDocuments() (int64, error) {
 	return count, nil
 }
 
+// scheduleExpiration schedules the next expiration of documents to occur, from the minimum expiration value in the bucket.
 func (bucket *Bucket) scheduleExpiration() {
-	if nextExp, err := bucket.NextExpiration(); err == nil && nextExp > 0 {
-		bucket.scheduleExpirationAtOrBefore(nextExp)
-	}
-}
-
-func (bucket *Bucket) scheduleExpirationAtOrBefore(exp uint32) {
-	if exp > 0 {
-		bucket.mutex.Lock()
-		defer bucket.mutex.Unlock()
-		if exp < *bucket.nextExp || *bucket.nextExp == 0 {
-			bucket.nextExp = &exp
-			dur := expDuration(exp)
-			if dur < 0 {
-				dur = 0
-			}
-			debug("EXP: Scheduling in %s", dur)
-			if bucket.expTimer == nil {
-				bucket.expTimer = time.AfterFunc(dur, bucket.doExpiration)
-			} else {
-				bucket.expTimer.Reset(dur)
-			}
-		}
+	if nextExp, err := bucket.nextExpiration(); err == nil && nextExp > 0 {
+		bucket.expManager._scheduleExpirationAtOrBefore(nextExp)
```

Comment:
It doesn't look like we're holding the expiryManager lock here - should this be `scheduleExpirationAtOrBefore`?

Reply:
We actually don't need to lock this because it gets called from two places:

- OpenBucket (we don't need lock because we are only initializing a single bucket)
- from doExpiration callback where we are already locked.

We do need `scheduleExpirationAtOrBefore` since it is used in `Collection.postNewEvent` on a write.